src/std/net/socks/client.ss

```diff
@@ -0,0 +1,423 @@
+  (socks-bind sock socks4-send-request socks4-recv-reply (and address (inet-address address))))
+
+(def (socks5-bind sock address)
+  (socks-bind sock socks5-send-request socks5-recv-reply (and address (inet-address address))))
+
+(defrule (socks-connect sock-in send-request recv-reply address arg ...)
+  (let (sock sock-in)
+    (with ([host . port] address)
+      (send-request sock #x01 host port arg ...)
+      (recv-reply sock))))
+
+(defrule (socks-bind sock-in send-request recv-reply maybe-address-in)
+  (let ((maybe-address maybe-address-in) (sock sock-in))
+    (match maybe-address
+      ([host . port]
+       (send-request sock #x02 host port))
+      (else
+       (using (sock :- StreamSocket)
+         (let (domain (sock.domain))
+           (cond
+            ((fx= domain AF_INET)
+             (send-request sock #x02 inaddr-any4 0))
+            ((fx= domain AF_INET6)
+             (send-request sock #x02 inaddr-any6 0))
+            (else
+             (error "unuspoorted socket domain" domain)))))))
+    (recv-reply sock)))
+
+(def (socks4-send-request sock cmd host port (4a? #f))
+  (using ((sock :- StreamSocket)
+          (pkt (open-buffered-writer #f 384) :- BufferedWriter))
+    (pkt.write-u8 #x04)                 ; VERSION
+    (pkt.write-u8 cmd)                  ; CMD
+    (pkt.write-u16 port)                ; PORT
+    (cond
+     ((and 4a? (string? host))
+      (pkt.write '#u8(0 0 0 1))         ; SOCKS4a bit
+      (pkt.write-u8 0)                  ; userid NULL terminator
+      (pkt.write-string host)           ; hostname
+      (pkt.write-u8 0))                  ; hostname NULL terminator
+     ((ip4-address? host)
+      (pkt.write host)                  ; host ip4
+      (pkt.write-u8 0))                 ; userid NULL terminator
+     (else
+      (raise-bad-argument socks4-send-request "bad host address" host)))
+    (let (msg (get-buffer-output-u8vector pkt))
+      (sock.send msg))))
+
+(def (socks4-recv-reply sock)
+  (using (sock :- StreamSocket)
+    (let (msg (make-u8vector 8 0))
+      (sock.recv msg)
+      (let (vn (u8vector-ref msg 0))
+        (unless (fx= vn 0)
+          (raise-io-error socks4-recv-reply "version mismatch" vn)))
+      (let (res (u8vector-ref msg 1))
+        (unless (fx= res 90)
+          (raise-io-error socks4-recv-reply "request rejected" res)))
+      (cons (subu8vector msg 4 8)
+            (fxior (fxarithmetic-shift (u8vector-ref msg 2) 8)
+                   (u8vector-ref msg 3))))))
+
+(def (socks5-send-request sock cmd host port)
+  (using ((sock :- StreamSocket)
+          (pkt (open-buffered-writer #f 384) :- BufferedWriter))
+    (pkt.write-u8 5)                    ; VER
+    (pkt.write-u8 cmd)                  ; CMD
+    (pkt.write-u8 0)                    ; RSV
+    (cond
+     ((string? host)
+      (pkt.write-u8 3) ; ATYP: DOMAINNAME
+      (let* ((fqdn (string->utf8 host))
+             (len (u8vector-length fqdn)))
+        (when (fx> len 255)
```

Comment:
Really, check-argument should allow more than one argument in its irritants, and then you'd use that instead of when... raise-bad-argument.

Reply:
this comes from the network, check-argument is inappropiate.